### PR TITLE
Update setup logic in model and datamodule

### PR DESCRIPTION
**Pull request by djdameln**
## 📝 Description

Minor change in how we check if `self._setup` should be called in `setup`. The change is repeated in both the datamodule and the anomaly module. This is mainly to ensure that we can use the modules in a standalone fashion, without/before being attached to a trainer. For example, we could use the following to quickly inspect the images in our datamodule.
```python
>>> from anomalib.data import MVTec
>>>
>>> datamodule = MVTec(image_size=(512, 512))
>>> datamodule.setup()
>>>
>>> sample = next(iter(datamodule.train_dataloader()))
>>> sample["image"].shape
torch.Size([32, 3, 512, 512])
```
Similarly, we could inspect the structure of the torch model of a given Lightning model after invoking `setup`:
```python
>>> from anomalib.models import Ganomaly
>>>
>>> model = Ganomaly()
>>> model.setup()
>>>
>>> model.model.generator.encoder1.input_layers
Sequential(
  (initial-conv-3-64): Conv2d(3, 64, kernel_size=(4, 4), stride=(2, 2), padding=(4, 4), bias=False)
  (initial-relu-64): LeakyReLU(negative_slope=0.2, inplace=True)
)
```

Without the current change, this would lead to problems downstream when the entrypoints of the engine are called.

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔨 Refactor (non-breaking change which refactors the code base)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](../docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/data/base/datamodule.py [hunk 1]
```diff
@@ -110,18 +111,22 @@ def __init__(
...
+
     def setup(self, stage: str | None = None) -> None:
         """Set up train, validation and test data.
 
         Args:
             stage: str | None:  Train/Val/Test stages.
                 Defaults to ``None``.
         """
-        if not self.is_setup:
+        if not any(hasattr(self, subset) for subset in ["train_data", "val_data", "test_data"]) or not self._is_setup:
```
Comment (samet-akcay) [inline src/anomalib/data/base/datamodule.py:123]: would it be possible to split this condition with an easier-to-read variable names?

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks, only a single comment...

File: src/anomalib/data/base/datamodule.py (hunk 1 above)
Comment (djdameln) [reply src/anomalib/data/base/datamodule.py]: done

**Review by samet-akcay — APPROVED** [review]
Great, thanks!

**Pull request merged**